### ISSUE TITLE
I modified the predicted trajectory using the Navila model, but the curves are not good. How did you solve the trajectory turning problem?

### ISSUE THREAD

**Issue by scuizhibin** [om-ai-lab/OpenTrackVLA#4]

**Comment by scuizhibin** [comment]
> Could you explain in more detail how you obtain the trajectory using NaVILA and what the resulting trajectory looks like?

Insert 32 placeholders (12800) into the input part of the LLM model, and extract the positions corresponding to the hidden_states placeholders as the input for trajectory prediction.

**Comment by scuizhibin** [comment]
按照占位符的方式，发现直线比较好，转弯的地方不行。


---Original---
From: &quot;Kyusong ***@***.***&gt;
Date: Tue, Dec 23, 2025 16:59 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [om-ai-lab/OpenTrackVLA] I modified the predicted trajectoryusing the Navila model, but the curves are not good. How did you solve thetrajectory turning problem? (Issue #4)


kyusonglee left a comment (om-ai-lab/OpenTrackVLA#4)
 
Could you explain in more detail how you obtain the trajectory using NaVILA and what the resulting trajectory looks like?
 
&mdash;

**Comment by kyusonglee** [comment]
You explained the feature extraction part. What I’m still wondering is how you actually generate the trajectory using NaVILA. Could you share the relevant code snippet? I can take a look.

**Comment by scuizhibin** [comment]
> You explained the feature extraction part. What I’m still wondering is how you actually generate the trajectory using NaVILA. Could you share the relevant code snippet? I can take a look.
 
llM 模型的训练部分 
hiden_state=torch.cat(hiden_state_list,dim=0)        
        if hiden_state.shape[1]!=32:
            print("hiden_state   -- hiden_state ",hiden_state.shape)
        if self.training and self.config.time_token_ids:
            outputs.loss = soft_cross_entropy(
                outputs.logits,
                new_labels,
                soft_tokens=self.config.time_token_ids,
                std=self.config.soft_ce_std,
            )
        
        loss_weight = calculate_loss_weight(new_labels)
        traj_l2_loss=0
        if self.traj_bool:
            if self.mm_traj.mm_projector_type=="dif":
               # print(" dif --dif ",self.mm_traj.mm_projector_type,hiden_state.shape, trajectory.shape)
                pred_result = self.mm_traj(hiden_state, trajectory)
                traj_loss_dict=self.mm_traj.compute_loss(pred_result,  traj_score)
                traj_l2_loss= traj_loss_dict['loss']
            else:
                out_traj= self.mm_traj(hiden_state)
                traj_l2_loss=self.l2_loss_fn(out_traj, trajectory)    
        else:
           traj_l2_loss=0
        outputs.loss = outputs.loss *  loss_weight *language_model_+traj_l2_loss

# ---traj_model 轨迹模型部分
class MultitrajProjector(PreTrainedModel):
    config_class = TrajProjectorConfig

    def __init__(self, mm_projector_cfg: TrajProjectorConfig, config: PretrainedConfig):
        super().__init__(mm_projector_cfg)

        self.mm_projector_type=mm_projector_cfg.mm_projector_type
        mm_input= mm_projector_cfg.mm_input
        mm_hidden_size=mm_projector_cfg.mm_hidden_size
        self.wiget_loss=mm_projector_cfg.wiget_loss
        # print("mm_input    -- mm_input ",mm_input)
        # print("mm_hidden_size    --  mm_hidden_size ",mm_hidden_size)
        self.queries = nn.Parameter(torch.randn(16,mm_input))
        self.attn = nn.MultiheadAttention(mm_input, num_heads=8)
        
 
        self.layers = nn.Sequential(
                nn.Linear(mm_input,mm_hidden_size),
                nn.LayerNorm(mm_hidden_size),
                nn.Linear(mm_hidden_size, mm_hidden_size),
                #nn.GELU(),
                nn.LayerNorm(mm_hidden_size),
                nn.Linear(mm_hidden_size, 3),
            )

    def _init_weights(self, module):
        # 检查当前模块是否是全连接层 (nn.Linear)
        if isinstance(module, nn.Linear):
            # 将权重 (weight) 初始化为零
            nn.init.kaiming_uniform_(module.weight, mode='fan_in', nonlinearity='relu')
            # 检查偏置项 (bias) 是否存在且需要初始化
            if module.bias is not None:
                # 将偏置项 (bias) 初始化为零
                nn.init.constant_(module.bias, 0)  

    def forward(self, hidden_states, *args, **kwargs):
        q = self.queries.unsqueeze(1).repeat(1, hidden_states.size(0), 1)
        k = v = hidden_states.transpose(0, 1)
        out, _ = self.attn(q, k, v) 
        x=out.transpose(0, 1) #256 
    
        x= self.layers(x)
        
        return x

**Comment by kyusonglee** [comment]
It looks like you’re training a separate mm_traj head that predicts trajectories from the hidden states. That means you must already have a ground-truth trajectory with a shape like [B, 16, 3].
However, in the NaVILA paper, the supervision is based on step-by-step actions (distance and angle), not on predicting 3D waypoint trajectories.
Could you explain where this ground-truth trajectory comes from and how you converted NaVILA’s outputs into it? It would be good to first check whether the ground-truth trajectory itself looks reasonable.

**Comment by scuizhibin** [comment]
> It looks like you’re training a separate mm_traj head that predicts trajectories from the hidden states. That means you must already have a ground-truth trajectory with a shape like [B, 16, 3]. However, in the NaVILA paper, the supervision is based on step-by-step actions (distance and angle), not on predicting 3D waypoint trajectories. Could you explain where this ground-truth trajectory comes from and how you converted NaVILA’s outputs into it? It would be good to first check whether the ground-truth trajectory itself looks reasonable.
We actually collected some motion data from the robot dog, specifically 2Hz data. Using four frames of images, we predicted the trajectory for the next two frames, and then interpolated the actual trajectory to obtain a 16*2 dimension true trajectory. This is equivalent to the trajectory in BEV (Behavioral Analysis).

**Comment by kyusonglee** [comment]
Sounds good. If you want to predict trajectories, you can try OpenTrackVLA. It should work.
